pandas/core/groupby/groupby.py

```diff
@@ -3484,6 +3484,47 @@ def shift(self, periods=1, freq=None, axis=0, fill_value=None):
         )
         return res
 
+    @final
+    @Substitution(name="groupby")
+    @Appender(_common_see_also)
+    def diff(self, periods=1, axis=0):
+        """
+        First discrete difference of element.
+
+        Calculates the difference of each element compared with another
+        element in the group (default is element in previous row).
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Periods to shift for calculating difference, accepts negative values.
+        axis : axis to shift, default 0
+            Take difference over rows (0) or columns (1).
+
+        Returns
+        -------
+        Series or DataFrame
+            First differences.
+        """
+        if axis != 0:
+            return self.apply(lambda x: x.diff(periods=periods, axis=axis))
+
+        obj = self._obj_with_exclusions
+        shifted = self.shift(periods=periods, axis=axis)
+
+        # GH45562 - to retain existing behavior and match behavior of Series.diff(),
+        # int8 and int16 are coerced to float32 rather than float64.
+        dtypes_to_f32 = ["int8", "int16"]
+        if obj.ndim == 1:
+            if obj.dtype in dtypes_to_f32:
+                shifted = shifted.astype("float32")
+        else:
+            mask = obj.dtypes.astype(str).isin(dtypes_to_f32).values
```

Comment:
umm what is this for?

Reply:
the existing groupby diff behavior upcast int8 and int16 to float32, not float64. However, groupby shift upcasts to float64. See #45562 for more discussion.

Reply:
its fine to break it here (and change the tests) upcatsing to float32 is weird

Reply:
Ok, will do. That means groupby.diff will upcast differently vs Series.diff, but that can be discussed in #45562.

Reply:
no you would fix that too

Reply:
oh i see we did that because of some cython issue. ok then let's preserve it (you will need an explict test for this ) and i don't think what you are doing is actually correct for frames

Reply:
thanks, fixed the type coercion and added an explicit test.